src/modules/job-manager/plugins/perilog.c

```diff
@@ -0,0 +1,513 @@
+                        proc->prolog ? "prolog": "epilog",
+                        stream);
+        return;
+    }
+    if (len) {
+        int level = LOG_INFO;
+        if (strcmp (stream, "stderr") == 0)
+            level = LOG_ERR;
+        flux_log (h,
+                  level,
+                  "%ju: %s: %s: %s",
+                  (uintmax_t) proc->id,
+                  proc->prolog ? "prolog" : "epilog",
+                  stream,
+                  s);
+    }
+}
+
+static int run_command (flux_plugin_t *p,
+                        flux_plugin_arg_t *args,
+                        int prolog,
+                        flux_cmd_t *cmd)
+{
+    flux_t *h = flux_jobtap_get_flux (p);
+    struct perilog_proc *proc;
+    flux_jobid_t id;
+    uint32_t userid;
+    flux_subprocess_t *sp = NULL;
+    flux_subprocess_ops_t ops = {
+        .on_completion = completion_cb,
+        .on_state_change = state_cb,
+        .on_stdout = io_cb,
+        .on_stderr = io_cb
+    };
+    char path[PATH_MAX + 1];
+    char jobid[128];
+    int rc = -1;
+
+    if (flux_plugin_arg_unpack (args,
+                                FLUX_PLUGIN_ARG_IN,
+                                "{s:I s:i}",
+                                "id", &id,
+                                "userid", &userid) < 0) {
+        flux_log_error (h, "flux_plugin_arg_unpack");
+        return -1;
+    }
+
+    if (flux_job_id_encode (id, "f58", jobid, sizeof (jobid)) < 0) {
+        flux_log_error (h, "flux_job_id_encode");
+        return -1;
+    }
+
+    if (flux_cmd_setcwd (cmd, getcwd (path, sizeof (path))) < 0
+        || flux_cmd_setenvf (cmd, 1, "FLUX_JOB_ID", "%s", jobid) < 0
+        || flux_cmd_setenvf (cmd, 1, "FLUX_JOB_USERID", "%u", userid) < 0) {
+        flux_log_error (h, "%s: flux_cmd_create", prolog ? "prolog" : "epilog");
+        goto done;
+    }
+
+    if (!(sp = flux_rexec (h, 0, 0, cmd, &ops))) {
+        flux_log_error (h, "%s: flux_rexec", prolog ? "prolog" : "epilog");
+        goto done;
+    }
+
+    if (!(proc = perilog_proc_create (p, id, prolog, sp))) {
+        flux_log_error (h, "%s: proc_create", prolog ? "prolog" : "epilog");
+        goto done;
+    }
+
+    if (flux_subprocess_aux_set (sp, "perilog_proc", proc, NULL) < 0) {
+        flux_log_error (h, "%s: flux_subprocess_aux_set",
+                        prolog ? "prolog" : "epilog");
+        goto done;
+    }
+
+    if (flux_jobtap_job_aux_set (p,
+                                 FLUX_JOBTAP_CURRENT_JOB,
+                                 "perilog_proc",
+                                 proc,
+                                 NULL) < 0) {
+        flux_log_error (h, "%s: flux_jobtap_job_aux_set",
+                        prolog ? "prolog" : "epilog");
+        goto done;
+    }
+
+    if (prolog) {
+        if (flux_jobtap_job_subscribe (p, FLUX_JOBTAP_CURRENT_JOB) < 0)
+            flux_log_error (h, "prolog: flux_jobtap_job_subscribe");
+    }
+
+    rc = 0;
+done:
```

Comment:
Remove label and just return -1 from each error path if there is no cleanup to do?

(is sp leaked on `perilog_proc_create()` failure?)

Reply:
> (is sp leaked on perilog_proc_create() failure?)

Yes, thanks!